File: p8/p8.js
const fs = require('fs');

const FILENAME = 'p8/p8_input.txt';

const getData = (fn) =>
  fs
    .readFileSync(fn)
    .toString()
    .split('\n')
    .map((x) => x.split('').map((x) => parseInt(x)));

const returnVisible = (xIdx, yIdx, arr) => {
  const pos = arr[xIdx][yIdx];
  const row = arr[xIdx];
  const col = arr.map((y) => y[yIdx]);

  if (
    xIdx === 0 ||
    xIdx === row.length - 1 ||
    yIdx === 0 ||
    yIdx === col.length - 1
  ) {
    return true;
  }

  const left = row.slice(0, yIdx);
  const right = row.slice(yIdx + 1);
  const up = col.slice(0, xIdx);
  const down = col.slice(xIdx + 1);

  const maxes = [left, right, up, down].map((x) =>
    x.length ? Math.max(...x) : 0
  );
  const min = Math.min(...maxes);
  if (pos > min) {
    return true;
  }
  return false;
};

const getNumLower = (pos, l) => {
  total = 0;
  while (l.length) {
    const t = l.shift();
    total += 1;
    if (t >= pos) {
      break;
    }
  }
  return total;
};

const getVisibility = (xIdx, yIdx, arr) => {
  const pos = arr[xIdx][yIdx];
  const row = arr[xIdx];
  const col = arr.map((y) => y[yIdx]);

  const left = row.slice(0, yIdx).reverse();
  const right = row.slice(yIdx + 1);
  const up = col.slice(0, xIdx).reverse();
  const down = col.slice(xIdx + 1);

  const total = [left, right, up, down]
    .filter((x) => x.length)
    .map((x) => getNumLower(pos, x))
    .reduce((a, b) => a * b, 1);
  return total;
};

const findTotal = (arr) => {
  let count = 0;
  arr.forEach((x, xIdx) => {
    x.forEach((y, yIdx) => {
      if (returnVisible(xIdx, yIdx, arr)) {
        count += 1;
      }
    });
  });
  return count;
};

const findTotalP2 = (arr) => {
  let maxVis = 0;
  arr.forEach((x, xIdx) => {
    x.forEach((y, yIdx) => {
      total = getVisibility(xIdx, yIdx, arr);
      if (total > maxVis) {
        maxVis = total;
      }
    });
  });
  return maxVis;
};

function main() {
  const data = getData(FILENAME);
  const total = findTotal(data);
  const totalP2 = findTotalP2(data);
  console.log(total, totalP2);
}

main();

File: p8/p8.py
import csv
import copy
import pprint
pp = pprint.PrettyPrinter(indent=4)
import numpy as np

FILENAME = "p8/p8_input.txt"

def get_data():
    output = []
    with open(FILENAME, 'r') as fd:
        reader = csv.reader(fd)
        for row in reader:
            output.append([int(i) for i in list(next(iter(row),None))])
    return output


def return_visible(ridx, cidx, arr):
    pos = arr[ridx,cidx]
    row = arr[ridx,:]
    col = arr[:,cidx]

    left = row[:cidx]
    right = row[cidx+1:]
    up = col[:ridx]
    down = col[ridx+1:]

    # edge 
    if ridx == 0 or ridx == len(row)-1:
        return True
    if cidx == 0 or cidx == len(col)-1:
        return True
    
    #interior but visible
    max_each_dir = [max(left), max(right), max(up), max(down)] 
    if pos > min(max_each_dir):
        return True
    return False


def get_num_lower(pos,l):
    l = list(l)
    total = 0
    while l:
        t = l.pop(0)
        total += 1
        if t >= pos:
            break
    return total

def get_visibility(ridx, cidx, arr):
    
    pos = arr[ridx,cidx]
    row = arr[ridx,:]
    col = arr[:,cidx]

    left = np.flip(row[:cidx])
    right = row[cidx+1:]
    up = np.flip(col[:ridx])
    down = col[ridx+1:]
    total = 1
    directions = [get_num_lower(pos,direction) for direction in [left,right,up,down]]
    for d in directions:
        total *= d
    return total
    

def process(arr):
    count = 0 
    mat = np.array(arr)
    for ridx,r in enumerate(mat):
        for cidx,c in enumerate(r):    
            if return_visible(ridx,cidx,mat):
                count += 1
    return count

def process_p2(arr):
    max_vis = 0
    max_coords = [None,None]
    mat = np.array(arr)
    print(mat)
    for ridx,r in enumerate(mat):
        for cidx,c in enumerate(r):    
            vis = get_visibility(ridx,cidx,mat)
            if vis > max_vis:
                max_vis = vis
                max_coords = [ridx,cidx]
    return max_vis

def main():
    arr = get_data()
    print(len(arr[0]))
    total = process(arr)
    total_p2 = process_p2(arr) 
    print(total, total_p2)

main()